src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,184 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using DnsClient;
+using DnsClient.Protocol;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    internal class DnsMonitor
+    {
+        #region static
+        private static string GetParentDomainName(string domainName)
+        {
+            var index = domainName.IndexOf('.');
+            return domainName.Substring(index + 1);
+        }
+        #endregion
+
+        // private fields
+        private readonly CancellationToken _cancellationToken;
+        private readonly IDnsMonitoringCluster _cluster;
+        private readonly LookupClient _lookupClient;
+        private readonly string _lookupDomainName;
+        private readonly string _parentDomainName;
+        private readonly string _query;
+        private bool _startReturnedNormally; // used for testing
+
+        // constructors
+        public DnsMonitor(IDnsMonitoringCluster cluster, string lookupDomainName, CancellationToken cancellationToken)
+        {
+            _cluster = Ensure.IsNotNull(cluster, nameof(cluster));
+            _lookupDomainName = Ensure.IsNotNull(lookupDomainName, nameof(lookupDomainName));
+            _cancellationToken = cancellationToken;
+            _lookupClient = new LookupClient();
+            _parentDomainName = GetParentDomainName(lookupDomainName);
+            _query = "_mongodb._tcp." + _lookupDomainName;
+        }
+
+        // public properties
+        public bool StartReturnedNormally => _startReturnedNormally;
+
+        // public methods
+        public void Start()
+        {
+            try
+            {
+                Monitor();
+            }
+            catch (OperationCanceledException)
+            {
+                // ignore OperationCanceledException
+            }
+            catch (Exception)
+            {
+                // TODO: log unexpected exceptions
+                throw;
+            }
+            _startReturnedNormally = true;
```

Comment:
This is the only way I could find to test that the DnsMonitor thread exited normally.

Before I added this some tests that were supposedly testing that the DnsMonitor thread exited when asked to were passing, but only because bugs in DnsMonitor were causing the thread to die due to an unhandled exception.